### PR TITLE
Fuzz binary search

### PR DESCRIPTION
**Pull request by jamii**
I wasn't really expecting to find anything, but a little more coverage can't hurt.

### CODE REVIEW

File: scripts/unique_fuzz_errors.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
```
Comment (sentientwaffle) [inline scripts/unique_fuzz_errors.sh]: Nit: rename file to `fuzz_unique_errors.sh` (as a convention, we use common prefixes to group files, types, functions, variables, etc).

File: src/lsm/binary_search.zig [hunk 2]
```diff
@@ -18,7 +18,7 @@ pub fn binary_search_values_raw(
     values: []const Value,
     key: Key,
 ) u32 {
-    assert(values.len > 0);
+    if (values.len == 0) return 0;
```
Comment (sentientwaffle) [inline src/lsm/binary_search.zig:27]: The assertion here is actually deliberate (albeit surprising). It forces the caller to specifically handle the empty case.

**Comment by jamii** [comment]
I somehow branched off jamii-fuzz-lsm instead of main. Will fix.

File: src/lsm/binary_search.zig (hunk 2 above)
Comment (sentientwaffle) [reply src/lsm/binary_search.zig:27]: Though I don't remember why :wink:. @jorangreef would know.

Comment (jamii) [reply src/lsm/binary_search.zig:27]: My instinct was that the caller always has to handle the case where result == values.len anyway, but maybe there are bad cases I haven't thought of yet.

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: The assertion is to document that within the `lsm` namespace we expect to only binary search values in a table, where tables always contain values. This means we don't require an optional return value. i.e. We can always return an index to a key.

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: The `result == values.len` case also shouldn't happen, because we first check the `key_min` and `key_max` of a table to see if the table is in range. We shouldn't be binary searching within a table if the result is going to be outside and to the right of the range.

Comment (sentientwaffle) [reply src/lsm/binary_search.zig:27]: We use binary search in other contexts in lsm besides searching tables (`SegmentedArray`, `ManifestLevel`).

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: Ah, good point. We should still always expect to be searching for `values.len > 0` there though?

Comment (sentientwaffle) [reply src/lsm/binary_search.zig:27]: When we are searching for an insertion index, we need to manually check for an empty array first, e.g.: https://github.com/tigerbeetledb/tigerbeetle/blob/29f4dfd758068e4c7d0b72d61917146bc735738f/src/lsm/manifest_level.zig#L580-L583

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: Would you prefer an optional result signature then for binary search to return null `if values.len == 0`?

Then we can assert on this by `.?` when searching tables.

Comment (sentientwaffle) [reply src/lsm/binary_search.zig:27]: :thinking: Would it be more consistent to return `0`, to match the case where the key is greater than all keys in the list and it returns `values.len`?

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: It should actually always return a valid index to the next greatest key. At least, that's the documentation there. That's how we used it also I think in the old ManifestLevel.

Comment (jamii) [reply src/lsm/binary_search.zig:27]: Returning 0 and exact=false does match the invariants. It's also correct for the case where we're trying to insert something into a sequence, regardless of whether the sequence is currently empty.

(If we don't put the assert back in, I should go and replace the usages that special-case empty slices)

Comment (jamii) [reply src/lsm/binary_search.zig:27]: We wrote concurrent and disagreeing comments :D

Comment (jamii) [reply src/lsm/binary_search.zig:27]: ~~...of which yours is correct, on further reading.~~

Comment (jamii) [reply src/lsm/binary_search.zig:27]: We currently have:

`.exact = index < values.len and compare_keys(key_from_value(&values[index]), key) == .eq,`

Which is misleading if we guarantee `index < values.len` already.

Comment (jamii) [reply src/lsm/binary_search.zig:27]: If we return the last index, rather than the end-of-the-array index in the case where the needle is greater than everything in the slice, then I think https://github.com/tigerbeetledb/tigerbeetle/blob/29f4dfd758068e4c7d0b72d61917146bc735738f/src/lsm/segmented_array.zig#L612 is unreachable? Potentially wrong?

Comment (jamii) [reply src/lsm/binary_search.zig:27]: I don't think there are any other cases that do anything with the index if exact == false.

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: Agreed with you on all three points. :D

We should clean this up—at present it reflects the evolution of how binary search was introduced, and it's not consistent at all.

Your interface would be perfect.

Comment (jamii) [reply src/lsm/binary_search.zig:27]: On stepping through, I think @jorangreef is right that the documentation and code match for search_raw. I should have read that closely before changing the behavior

I find the documented behavior surprising though and the code in segmented_array at least seems to assume the behavior that I expect for binary_search - to return the insertion point if there is no exact match.

Comment (jamii) [reply src/lsm/binary_search.zig:27]: ...github needs a typing indicator :D

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: @jorangreef is typing...

Comment (sentientwaffle) [reply src/lsm/binary_search.zig:27]: > I find the documented behavior surprising though and the code in segmented_array at least seems to assume the behavior that I expect for binary_search - to return the insertion point if there is no exact match.

Yeah, I didn't read the binary search docs closely enough when updating the segmented array :sweat_smile:.

Comment (jamii) [reply src/lsm/binary_search.zig:27]: Ok, todo for me next week:
* [x] Change the documentation and tests to match the usual behavior of binary search.
* [ ] Review all usages to make sure no correct behavior changes.
* [ ] Test if the current usage in segmented arrays is correct.
* [ ] Check the inlined search impl in segmented array

Comment (jorangreef) [reply src/lsm/binary_search.zig:27]: Let's fix it to be non-surprising!

Comment (sentientwaffle) [reply src/lsm/binary_search.zig:27]: Just for a bonus gotcha, segmented array also has an inlined implementation of binary search (in `search()`) to iterate over nodes, before it drops into `binary_search_values_raw()` to search within the found node. :grin:

Comment (jamii) [reply src/lsm/binary_search.zig:27]: I added a lot of config.verify -only asserts and tweaked the explict tests to cover edge cases. It seems that @sentientwaffle's expectations were accurate. So I didn't have to change the implementation, but I made the docs more explicit so there is no confusion.

**Comment by jamii** [comment]
@sentientwaffle I didn't end up changing the behavior of binary_search, only clarified the docs and expanded the tests. So I think this pr is good and I'll look at segmented array in a separate pr.

**Comment by sentientwaffle** [comment]
Sounds good! I'll review this today.

File: src/lsm/binary_search.zig [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
 // which should eliminate the current poor codegen of key_from_value/compare_keys.
 
-/// Returns the index of the key either exactly equal to the target key or, if there is no exact
-/// match, the next greatest key.
+/// Returns either the index of the first value equal to `key`,
+/// or if there is no such value then the index where `key` would be inserted
+/// ie returns `i` such that both:
+/// * i == values.len or key_from_value(values[i]) >= key
+/// * i == 0 or key_value_from(values[i-1]) < key
+///
```
Comment (sentientwaffle) [inline src/lsm/binary_search.zig]: I swapped the `or` order so that it aligned more, which makes it easier to spot similarities + differences at a glance.

```suggestion
/// Returns either the index of the first value equal to `key`,
/// or if there is no such value then the index where `key` would be inserted.
///
/// In other words, return `i` such that both:
/// * key_from_value(values[i])  >= key or i == values.len
/// * key_value_from(values[i-1]) < key or i == 0
///
```

File: src/lsm/binary_search.zig [hunk 4]
```diff
@@ -18,11 +23,23 @@ pub fn binary_search_values_raw(
     values: []const Value,
     key: Key,
 ) u32 {
-    assert(values.len > 0);
+    if (values.len == 0) return 0;
+
+    if (config.verify) {
+        // Input must be sorted by key.
+        for (values) |_, i| {
+            assert(i == 0 or compare_keys(key_from_value(&values[i - 1]), key_from_value(&values[i])) != .gt);
```
Comment (sentientwaffle) [inline src/lsm/binary_search.zig]: Codebase style is to wrap at 100 chars. Not sure what `zig fmt` will allow, but something like
```suggestion
            assert(i == 0 or
                compare_keys(key_from_value(&values[i - 1]), key_from_value(&values[i])) != .gt);
```

File: src/lsm/binary_search.zig [hunk 5]
```diff
@@ -164,6 +198,40 @@ const test_binary_search = struct {
             try std.testing.expectEqual(expect.exact, actual.exact);
         }
     }
+
+    fn random_search(random: std.rand.Random, iter: usize) !void {
+        const keys_count = @floatToInt(usize, @trunc(random.floatExp(f64) * @intToFloat(f64, iter)));
```
Comment (sentientwaffle) [inline src/lsm/binary_search.zig]: Could we extract this to a function (from here and the other fuzzers that use it) into a `test/fuzz.zig`? And include a `max` parameter to keep a sane limit for the unit tests' memory requirements. (It is fine if that is the only function there for now; we can fill it out over time).

If a fuzzer OOMs I want it to be because of a bug, not a bad roll of the RNG. :sweat_smile:

File: src/lsm/binary_search.zig (hunk 4 above)
Comment (jamii) [reply src/lsm/binary_search.zig]: Should the lint be running in CI?

```
zig run ./scripts/lint.zig -- $(rg --files src/ | grep '.zig') 2>&1 | grep -c 'line exceeds 100'
258
```

Comment (sentientwaffle) [reply src/lsm/binary_search.zig]: Not for now. Fixing all of those would cause a bunch of git conflicts with open branches right now. Maybe we could try to fix it gradually, and then when the number is much lower add it to CI.

**Comment by jamii** [comment]
@sentientwaffle I updated all the callers of binary_search, moved some of the random generation into test/fuzz and capped the size of all the random tests.

File: src/test/fuzz.zig [hunk 6]
```diff
@@ -0,0 +1,47 @@
+//! Utils functions for writing fuzzers.
+
+const std = @import("std");
+const assert = std.debug.assert;
+
+/// Returns an integer of type `T` with an exponential distribution of rate `avg`.
+/// Note: If you specify a very high rate then `std.math.maxInt(T)` may be over-represented.
+pub fn random_int_exp(random: std.rand.Random, comptime T: type, avg: T) T {
```
Comment (sentientwaffle) [inline src/test/fuzz.zig]: Let's be more precise: `random_int_exponential`.

File: src/test/fuzz.zig [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        total = 0;
+        for (biases) |bias| total += bias;
+        if (total != 0) break;
+    }
+    for (biases) |*bias| bias.* = bias.* / total;
+    return biases;
+}
+
+/// Generate a random `Enum`, given a list of probabilities for each field.
+pub fn random_enum_biased(
```
Comment (sentientwaffle) [inline src/test/fuzz.zig]: Rename to `random_biased_enum` to share a longer prefix with `random_biases`.

File: src/test/fuzz.zig [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        const info = @typeInfo(T);
+        assert(info == .Int);
+        assert(info.Int.signedness == .unsigned);
+    }
+    const exp = random.floatExp(f64) * @intToFloat(f64, avg);
+    return std.math.lossyCast(T, exp);
+}
+
+/// Return a list of probabilities for use with `random_enum_biased`.
+pub fn random_biases(
```
Comment (sentientwaffle) [inline src/test/fuzz.zig]: Would `random_biased_distribution` be an accurate name? To group with `random_biased_enum` below.

File: src/test/fuzz.zig (hunk 6 above)
Comment (jamii) [reply src/test/fuzz.zig]: Ok :)

I was copying Random.floatExp.

Comment (sentientwaffle) [reply src/test/fuzz.zig]: Ah, fair enough. We typically spell out full words when it isn't too painfully long, but since it is mirroring a stdlib API that's fine either way. Come to think of it we do have others like this, e.g. `div_ceil`.

File: src/test/fuzz.zig (hunk 8 above)
Comment (jamii) [reply src/test/fuzz.zig]: How about `random_enum_distribution` and `random_enum`?

Comment (sentientwaffle) [reply src/test/fuzz.zig]: Perfect, even better.

**Review by sentientwaffle — APPROVED** [review]
Looks good!
